File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode tail = head;
        int listSize = 1;
        int midPoint = 0;
        
        if(head.next == null)
            return head;
        
        while(tail.next != null){
            tail = tail.next;
            listSize++;
        }
        
         midPoint = listSize/2;
        
        while(midPoint > 0){
            head = head.next;
            midPoint--;
        }
        
        return head;
    }
}

File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        size = 0
        tail = head
        while tail != None:
            tail = tail.next
            size += 1
            
        mid = head
        midIndex = size // 2
        midIndexInt = midIndex
        for x in range(midIndexInt):
            mid = mid.next
            
        return mid
        